doc/man/man3/netpath.3.man
'\" rtp
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77
.\" =========================================================================
.\"
.\" @(#) doc/man/man3/netpath.3.man
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to copy, distribute and/or modify this manual under
.\" the terms of the GNU Free Documentation License, Version 1.3 or any later
.\" version published by the Free Software Foundation; with no Invariant
.\" Sections, no Front-Cover Texts, and no Back-Cover Texts.  A copy of the
.\" license is included in the section entitled "GNU Free Documentation
.\" License".
.\"
.\" Permission to use, copy and distribute this manual without modification,
.\" for any purpose and without fee or royalty is hereby granted, provided
.\" that both the above copyright notice and this permission notice appears
.\" in all copies and that the name of OpenSS7 Corporation not be used in
.\" advertising or publicity pertaining to distribution of this documentation
.\" or its contents without specific, written prior permission.  OpenSS7
.\" Corporation makes no representation about the suitability of this manual
.\" for any purpose.  It is provided "as is" without express or implied
.\" warranty.
.\"
.\" Permission is granted to process this file through groff and print the
.\" results, provided the printed document carries a copying permission
.\" notice identical to this one except for the removal of this paragraph
.\" (this paragraph not being relevant to the printed manual).
.\"
.\" OPENSS7 CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS MANUAL
.\" INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
.\" PARTICULAR PURPOSE, NON-INFRINGEMENT, OR TITLE; THAT THE CONTENTS OF THE
.\" DOCUMENT ARE SUITABLE FOR ANY PURPOSE, OR THAT THE IMPLEMENTATION OF SUCH
.\" CONTENTS WILL NOT INFRINGE ON ANY THIRD PARTY PATENTS, COPYRIGHTS,
.\" TRADEMARKS OR OTHER RIGHTS.  IN NO EVENT SHALL OPENSS7 CORPORATION BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES OR ANY
.\" DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
.\" IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH ANY USE OF THIS DOCUMENT OR THE PERFORMANCE
.\" OR IMPLEMENTATION OF THE CONTENTS THEREOF.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this manual
.\" page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from the
.\" use of the information contained herein.  The author(s) may not have
.\" taken the same level of care in the production of this manual, which is
.\" licensed free of charge, as they might when working professionally.  The
.\" author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by the
.\" source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
.\" behalf of the U.S. Government ("Government"), the following provisions
.\" apply to you.  If the Software is supplied by the Department of Defense
.\" ("DoD"), it is classified as "Commercial Computer Software" under
.\" paragraph 252.227-7014 of the DoD Supplement to the Federal Acquisition
.\" Regulations ("DFARS") (or any successor regulations) and the Government
.\" is acquiring only the license rights granted herein (the license rights
.\" customarily provided to non-Government users).  If the Software is
.\" supplied to any unit or agency of the Government other than DoD, it is
.\" classified as "Restricted Computer Software" and the Government's rights
.\" in the Software are defined in paragraph 52.227-19 of the Federal
.\" Acquisition Regulations ("FAR") (or any successor regulations) or, in the
.\" cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the
.\" FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database openss7.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
search-truncate 40
search-ignore CGIQOSTU
.R2
.so openss7.macros
.\"
.\"
.TH NETPATH 3 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Library Calls"
.\"
.\"
.SH NAME
.B netpath, setnetpath, getnetpath, endnetpath
\- Network Configuration path subroutines.
.\"
.\"
.SH SYNOPSIS
.PP
.B #include <netconfig.h>
.HP
.BI "void *setnetpath(void);"
.HP
.BI "struct netconfig *getnetpath(void *" handle );
.HP
.BI "int endnetpath(void *" handle );
.\"
.\"
.SH ARGUMENTS
.TP
.I handle
an opaque pointer returned by
.BR setnetpath ()
to be used as an argument to calls to
.BR getnetpath ().
.\"
.\"
.SH DESCRIPTION
.PP
.IR UNIX \(rg
has traditionally provided three components of the Network Services Library,
.BR libnsl ,
that are associated with Transport Provider Interface,
.BR tpi (7),
.IR STREAMS (9)
network pseudo-device drivers, and that are not currently part of the
.B Linux
Network Services Library.  These components are:
.IP 1. \w'1.\(em'u
Network Configuration database access, see
.BR netconfig (3).
.PD 0
.IP 2.
Network Directory and address mapping, see
.BR netdir (3).
.IP 3.
Network Path access, see
.BR netpath (3).
.PD
.PP
These are often refered to as the Network Selection facility of the Network
Services Library.
.PP
.BR setnetpath "(), " getnetpath "() and " endnetpath ()
provide a mechanism to sequentialy access entries in the Network Configuration database that
correspond to entries in the
.RB { NETPATH }
environment variable.  That is, the
.RB { NETPATH }
environment variable acts as a filter specification that filters network identifiers.
.TP
.BR setnetpath ()
returns a pointer to an opaque datastructure used as a handle to the
.BR getnetpath ()
subroutine.
.BR setnetpath ()
is used to initialize client access to the database entries and set the access point to the first
valid entry in the
.RB { NETPATH }
environment variable.  If the
.RB { NETPATH }
environment variable is not set, all of the entries in the
.BR netconfig (5s)
database that are marked
.RI \(lq visible \(rq
will be used instead.
.TP
.BR getnetpath (),
when called successively with the
.I handle
returned by the call to
.BR setnetpath (),
returns each valid Network Configuration database entry that appears in the
.RB { NETPATH }
environment variable starting with the first.  If the
.RB { NETPATH }
environment variable is not set, all of the database entries that are marked
.RI \(lq visible \(rq
will be used instead.
Once the last entry is returned,
.BR getnetpath ()
will subsequently return
.BR NULL .
.TP
.BR endnetpath ()
releases client access to the database and frees for reuse any resources that were held by
.BR setnetpath ()
for use by
.BR getnetpath ().
.BR endnetpath ()
must only be called with a
.I handle
returned from a previous and corresponding call to
.BR setnetpath ().
.BR endnetpath ()
must be called for each call to
.BR setnetpath ()
with the corresponding
.I handle
pointer as the argument.
.PP
.BR nc_perror (3)
and
.BR nc_sperror (3)
provide a mechanism for providing diagnostic messages upon failure.  These subroutines are used for
both
.BR netconfig (3)
subroutines as well as the
.B netpath
subroutines described here.
.\"
.\"
.SS "NETPATH FORMAT"
.PP
The format of the
.RB { NETPATH }
environment variable is a sequence of zero or more network identification strings as described in
.BR netconfig (5s),
separated by colons.
If the
.RB { NETPATH }
environment variable is not set, these subroutines act as though the
.RB { NETPATH }
environment variable consisted of each
.RI \(lq visible \(rq
network identification string entry in the
.BR netconfig (5s)
database, in the order they occur in the database, separated by colons.
.\"
.\"
.SS "DATABASE FORMAT"
.PP
The
.BR netconfig
structure is defined in
.BR netconfig (3)
and
.BR netconfig (5s).
.\"
.\"
.SH NOTICES
.PP
.BR nc_perror (3)
and
.BR nc_sperror (3)
must be called immediately after the failed network configuration subroutine fails.  Otherwise, the
error condition may be lost.
.\"
.\"
.SH "RETURN VALUE"
.PP
Upon successful completion,
.BR setnetpath ()
returns an opaque pointer for use in the
.BR getnetpath ()
call.  Otherwise,
.B NULL
is returned and
.IR nc_error (3)
is set to indicate the error.
.PP
Upon successful completion, 
.BR getnetpath ()
return a pointer to a
.BR netconfig (3)
structure.  Otherwise, it returns
.B NULL
to indicate failure.
.PP
Upon successful completion,
.BR endnetpath ()
returns 0.  Otherwise, -1 is
returned and
.IR nc_error (3)
is set to indicate the error.
.\"
.\"
.SH "ERROR HANDLING"
.PP
When an error occurs, an error message can be printed to standard error or retrieved using
.BR nc_perror (3)
or
.BR nc_sperror (3).
.\"
.\"
.SH ERRORS
.PP
The
.BR setnetpath ()
subroutine will faile, and return a
.B NULL
pointer, if: an error was encountered opening the Network Configuration database,
.BR /etc/netconfig ,
see
.BR netconfig (5s).
.PP
The
.BR getnetpath ()
subroutine will fail, and return a
.B NULL
pointer, if:
no etnries exist in the Network Configuration database corresponding to the
.RB { NETPATH }
environment variable; or, the last entry in the Network Configuration database associated with the
.RB { NETPATH }
environment variable has already been read with a previous call to
.BR getnetpath ()
and there are no more entries in the Network Configuration database corresponding to the
.RB { NETPATH }
environment variable.
.PP
The
.BR endnetpath ()
subroutine will fail if: the passed in
.I handle
argument is invalid; or,
.BR endnetpath ()
has already been called for the
.I handle
argument.
.PP
The following errors can be returned:
.TP \w'[\fBNC_NOMOREENTRIES\fR]\(em'u
.RB [ NC_NOERROR ]
.RI \(lq "no error" \(rq
.IP
The last network selection function called was successful or no network selection function has yet
been called.
.TP
.RB [ NC_NOMEM ]
.RI \(lq "out of memory" \(rq
.IP
There was not sufficient memory for the last network selection function called to complete.
.TP
.RB [ NC_NOSET ]
.RI \(lq "routine called before calling setnetpath() or setnetconfig()" \(rq
.IP
The last call to
.BR endnetpath ()
or
.BR getnetpath (3)
was called (incorrectly) before calling
.BR setnetpath (3).
.TP
.RB [ NC_OPENFAIL ]
.RI \(lq "cannot open /etc/netconfig" \(rq
.IP
The last call to a network selection function,
.BR setnetpath (3),
could not open the
.B /etc/netconfig
file.
.TP
.RB [ NC_BADLINE ]
.RI \(lq "error in /etc/netconfig : field %d of line %d" \(rq
.IP
The last call to a network selection function detected a syntax error in the
.B /etc/netconfig
file at field
.I %d
of line
.IR %d .
.TP
.RB [ NC_NOTFOUND ]
.RI \(lq "netid not found in /etc/netconfig" \(rq
.IP
The
.I netid
argument to the last call to
.BR getnetconfigent (3)
was not found in
.BR /etc/netconfig .
.TP
.RB [ NC_NOMOREENTRIES ]
.RI \(lq "no more entries in /etc/netconfig" \(rq
.IP
The last call to network selection function
.BR getnetpath (3)
exhausted the entries in the
.B /etc/netconfig
file.
.TP
.RB [ ??? ]
.RI \(lq "unknown error" \(rq
.IP
An internal error occurred in the
.BR libxnsl (3)
library.
.\"
.\"
.\".SH DEVICES
.\".PP
.\"
.\"
.SH FILES
.PP
.TP
.B /etc/netconfig
The Network Configuration database file.  See
.BR netconfig (5s).
.TP
.B /usr/lib/libxnsl.so
The X/Open Network Services Library shared object library.
.TP
.B /usr/lib/libxnsl.a
The X/Open Network Services Library static library.
.\"
.\"
.SH ENVIRONMENT
.PP
.TP
.RB { NETPATH }
This environment variable contains the network path.
.\"
.\"
.SH "SEE ALSO"
.PP
.BR environ (7),
.BR netconfig (5s),
.BR netconfig (3),
.BR nc_perror (3),
.BR nc_sperror (3),
.BR netdir (3),
.IR "RFC 1833" .
.[
rfc1833
.]
.\"
.\"
.SH BUGS
.PP
.BR setnetpath (),
.BR getnetpath "() and "
.BR endnetpath ()
had no known bugs.
.\"
.\"
.SH COMPATIBILITY
.PP
These network configuraiton subroutines are compatible with
.I "SVR 4.2"
.[
svr42
.]
and systems based on
.IR "SVR 4.2" ,
.[
svr42
.]
including
.IR AIX \(rg,
.[
aixdoc
.]
.IR HP-UX \(rg,
.[
hpuxdoc
.]
.IR Solaris \(rg
.[
soldoc
.]
and
.IR UnixWare \(rg,
.[
uw7doc
.]
with the following compatibility considerations:
.IP \(em \w'\(em\(em'u
.IR Solaris \(rg
.[
soldoc
.]
and
.IR UnixWare \(rg
.[
uw7doc
.]
describe the
.BR nc_sperror ()
subroutine as returning a pointer to thread-specific data and that the buffer
could change at any point.  While the error value is saved as thread-specific
data, this implementation of
.BR nc_sperror ()
returns a pointer to a shared read-only string that is valid for the duration
of the calling process.
.IP \(em
.B @PACKAGE_TITLE@
stores the error condition in
.IR nc_error (3).
.IP \(em
Diagnostic message strings are not documented and are incompatible.
.IP \(em
Binary compatibility would only be coincidental.
.IP \(em
On
.IR UNIX \(rg
systems, these utility subroutines are normally part of the Network Services
Library,
.BR libnsl ;
however,
.B Linux
already has a library of that name (which unfortunately is missing these
components), and, therefore,
.B "@PACKAGE_TITLE@"
renames the library to
.BR libxnsl (3).
.PP
See
.BR libxnsl (3)
for additional compatibility information.
.\"
.\"
.SH CONFORMANCE
.PP
.IR "SVR 4.2" .
.[
svr42
.]
.\"
.\"
.SH HISTORY
.PP
The network configuration subroutines first appeared in
.IR "SVR 4.2" .
.[
svr42
.]
These subroutines are related to the
.IR UNIX \(rg
.BR STREAMS (9)
implementation of network transport providers.
.\"
.\"
.[
$LIST$
.]
.TI


NEWS
OpenSS7 -- history of user visible changes.  2015-01-15
$Id: NEWS,v 1.1.2.25 2011-05-10 13:45:30 brian Exp $
Copyright (c) 2008-2015  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
'OpenSS7' package.

Major changes for release openss7-1.1.7.20150115
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Source code release on GitHub.

Major changes for release openss7-1.1.7.20131209
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build corrections for RHEL 6.4.

   - Scripts for country boundary, shoreline and rivers database
     preparation.

   - Database for ANSI T1.101 point code assignments.

   - Suppress extransous /proc/1/comm error messages when testing for
     systemd.

   - Handle PUIAS to Springdale rename for PUIAS distribution.

Major changes for release openss7-1.1.7.20131123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Correction to ca-cert handling in RHEL 6 install scriptlet.

   - RHEL 6 build changes.

   - Set permissions correctly when applying kernel module patch.

   - Do not duplicate certs when including ca-cert.

   - Avoid conflicting shelll variable in RPM install scriplet.

   - Remove /proc/1/comm checks for systemd in module loader.

Major changes for release openss7-1.1.7.20130209
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Corrections to RPM spec file.

Major changes for release openss7-1.1.7.20130129
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Compensate for RHEL6 lack of c_rehash.

   - Do not create sysfs nodes: they do not work properly.

   - Use no foreground flag in strace.service file.

   - More error traces for strerr logger.

   - Remove static device names (the do not work either).

   - Changes to support weak kernel modules.

   - Clean up weak module support.

Major changes for release openss7-1.1.7.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Better Archlinux install script support.

Major changes for release openss7-1.1.6.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Remove use of kill_litter_super().

Major changes for release openss7-1.1.3.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Updates and corrections to weak kernel module builds.

   - Repo support and corrections for Archlinux packages.

Major changes for release openss7-1.1.3.20130123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Additional support for Archlinux.

   - Add udev rules for strace and strerr.

   - Add system unit files for strace and strerr.

   - Add repo support for Archlinux packages.

   - Build Archlinux packages better.

Major changes for release openss7-1.1.3.20130121
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build support for SLES 11.

   - Handle certificates better.

   - Get configuration files in the right places for SLES 11.

   - Support in RPM spec file for certificates on SLES 11.

   - Create a new OpenSS7 certificate.

   - Repository support for SLES 11.

   - RPM spec file reworked for new RPM release.

   - Better certificate handling for RHEL.

   - Additional build support for RHEL 6 with certificates.


strsock/src/lib/netpath.c
/*****************************************************************************

 @(#) $RCSfile$ $Name$($Revision$) $Date$

 -----------------------------------------------------------------------------

 Copyright (c) 2001-2006  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software; you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free Software
 Foundation; version 2 of the License.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 details.

 You should have received a copy of the GNU General Public License along with
 this program; if not, write to the Free Software Foundation, Inc., 675 Mass
 Ave, Cambridge, MA 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified $Date$ by $Author$

 -----------------------------------------------------------------------------

 $Log$
 *****************************************************************************/

#ident "@(#) $RCSfile$ $Name$($Revision$) $Date$"

static char const ident[] = "$RCSfile$ $Name$($Revision$) $Date$";

#define _XOPEN_SOURCE 600
#define _REENTRANT
#define _THREAD_SAFE

#if 0
#define __USE_UNIX98
#define __USE_XOPEN2K
#define __USE_GNU
#endif

#define NEED_T_USCALAR_T 1

#include <stdlib.h>

#include "gettext.h"
#include <sys/types.h>
#include <sys/stat.h>
#ifdef HAVE_SYS_IOCTL_H
# include <sys/ioctl.h>
#endif
#include <sys/stropts.h>
#include <sys/poll.h>
#include <fcntl.h>

#ifdef HAVE_INTTYPES_H
# include <inttypes.h>
#else
# ifdef HAVE_STDINT_H
#  include <stdint.h>
# endif
#endif

#ifndef __EXCEPTIONS
#define __EXCEPTIONS 1
#endif

#include <unistd.h>
#include <errno.h>
#include <stddef.h>
#include <stdio.h>
#include <string.h>
#include <stropts.h>
#include <pthread.h>
#include <linux/limits.h>
#include <values.h>

#ifndef __P
#define __P(__prototype) __prototype
#endif

#include <sys/sockmod.h>
#include <sys/socksys.h>
#include <sys/socket.h>
//#include <tihdr.h>
//#include <timod.h>

#if defined __i386__ || defined __x86_64__ || defined __k8__
#define fastcall __attribute__((__regparm__(3)))
#else
#define fastcall
#endif

#define __hot __attribute__((section(".text.hot")))
#define __unlikely __attribute__((section(".text.unlikely")))

#if __GNUC__ < 3
#define inline static inline fastcall __hot
#define noinline extern fastcall __unlikely
#else
#define inline static inline __attribute__((always_inline)) fastcall __hot
#define noinline static __attribute__((noinline)) fastcall __unlikely
#endif

#define likely(x) __builtin_expect(!!(x), 1)
#define unlikely(x) __builtin_expect(!!(x), 0)

#undef min
#define min(a, b) (a < b ? a : b)

/**
 * @fn struct netconfig *getnetpath(void *handle);
 * @brief Get the next entry associated with handle.
 * @param handle a pointer to the handle returned by setnetpath().
 *
 * This function returns a pointer to the netconfig database entry corresponding to the first valid
 * NETPATH component.  The netconfig entry is formattted as a struct netconfig.  On each subsequenc
 * all, this function returns a pointer to the netconfig entry that corresponds to the next valud
 * NETPATH component.  This function can thus be used to search the netconfig database for all
 * networks included in the NETPATH variable.  When NETPATH has been exhausted, this function
 * returns NULL.
 *
 * This function silently ignores invalid NETPATH components.  A NETPATH component is invalid if
 * there is no corresponding entry in the netconfig database.
 *
 * If the NETPATH environment variable is unset, this function bethaves as if NETPATH were set to
 * the sequence of "default" or "visible" networks in the netconfig database, in the order in which
 * they are listed.
 *
 * When first called, this function returns a pointer to the netconfig database entry corresponding
 * to the first valid NETPATH component.  When NETPATH has been exhausted, it returns NULL.
 */
struct netconfig *
__nsl_getnetpath(void *handle)
{
	errno = EOPNOTSUPP;
	return (NULL);
}

struct netconfig *getnetpath(void *handle)
    __attribute__ ((weak, alias("__nsl_getnetpath")));

/**
 * @fn void *setnetpath(void);
 * @brief Get a handle for network configuration database.
 *
 * A call to this function binds to or rewinds NETPATH.  This function must be called before the
 * first call to getnetpath() and may be called at any other time.  It returns a handle that is used
 * by getnetpath().
 *
 * This function returns a handle that is used by getnetpath().  In case of an error, this function
 * returns NULL.  nc_perror() or nc_sperror() can be used to print out the reason for failure.  See
 * getnetconfig().
 */
void *
__nsl_setnetpath(void)
{
	errno = EOPNOTSUPP;
	return (NULL);
}

void *setnetpath(void)
    __attribute__ ((weak, alias("__nsl_setnetpath")));

/**
 * @fn int endnetpath(void *handle);
 * @brief Free netpath resources associated with handle.
 * @param handle a pointer to the handle returned by setnetpath().
 *
 * This function may be called to unbind from NETPATH when processing is complete, releasing
 * resources for reuse.  Programmers should be aware, however, that this function frees all memory
 * allocated by getnetpath() for the struct netconfig data structure.
 *
 * This function returns 0 on success and -1 on failure (for example, if setnetpath() was not called
 * previously).
 *
 */
int
__nsl_endnetpath(void *handle)
{
	errno = EOPNOTSUPP;
	return (-1);
}

int endnetpath(void *handle)
    __attribute__ ((weak, alias("__nsl_endnetpath")));
